### PR TITLE
Fix fundemental BoltDB backend correctness issues.

### LINKED ISSUE #3717: libs/db/bolt: Batch.Delete implementation is incorrect
**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source):

`develop`

**What happened**:

Manual code review of the BoltDB backend.

**Anything else we need to know**:

> One of these things is not like the others
> One of these things just doesn't belong
> Can you tell which thing is not like the others
> By the time I finish my song?

  * https://github.com/tendermint/tendermint/blob/96e132b4b0bd234bd3405c1909cdd40f19d2d259/libs/db/boltdb.go#L183
 * https://github.com/tendermint/tendermint/blob/96e132b4b0bd234bd3405c1909cdd40f19d2d259/libs/db/c_level_db.go#L169
 * https://github.com/tendermint/tendermint/blob/96e132b4b0bd234bd3405c1909cdd40f19d2d259/libs/db/go_level_db.go#L164
 * https://github.com/tendermint/tendermint/blob/96e132b4b0bd234bd3405c1909cdd40f19d2d259/libs/db/mem_batch.go#L35

### PR DESCRIPTION
**Pull request by Yawning**
* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [x] Updated CHANGELOG_PENDING.md

These changes fix #3717 and #3718.

### CODE REVIEW

File: libs/db/boltdb.go [hunk 1]
```diff
@@ -149,55 +151,45 @@ func (bdb *BoltDB) Stats() map[string]string {
...
+	bdb.ops = append(bdb.ops, operation{opTypeDelete, key, nil})
 }
 
 // NOTE: the operation is synchronous (see BoltDB for reasons)
 func (bdb *boltDBBatch) Write() {
 	err := bdb.db.db.Batch(func(tx *bbolt.Tx) error {
 		b := tx.Bucket(bucket)
-		for _, elem := range bdb.buffer {
-			if putErr := b.Put(elem.k, elem.v); putErr != nil {
-				return putErr
```
Comment (Liamsi) [inline libs/db/boltdb.go]: Ouch. How didn't we notice this during review 🤦‍♂ Thanks!

File: CHANGELOG_PENDING.md [hunk 2]
```diff
@@ -28,3 +28,5 @@
 * [rpc] [\#3686](https://github.com/tendermint/tendermint/pull/3686) `HTTPClient#Call` returns wrapped errors, so a caller could use `errors.Cause` to retrieve an error code.
 
 ### BUG FIXES:
+- [libs/db] Fixed the BoltDB backend's Batch.Delete implementation
```
Comment (Liamsi) [inline CHANGELOG_PENDING.md]:
```suggestion
- [libs/db] Fixed the BoltDB backend's Batch.Delete implementation (@Yawning)
```

File: CHANGELOG_PENDING.md [hunk 3]
```diff
@@ -28,3 +28,5 @@
...
+- [libs/db] Fixed the BoltDB backend's Get and Iterator implementation
```
Comment (Liamsi) [inline CHANGELOG_PENDING.md]:
```suggestion
- [libs/db] Fixed the BoltDB backend's Get and Iterator implementation (@Yawning)
```

File: CHANGELOG_PENDING.md (hunk 2 above)
Comment (Yawning) [reply CHANGELOG_PENDING.md]: Fixed in the force-pushed branch.

File: CHANGELOG_PENDING.md (hunk 3 above)
Comment (Yawning) [reply CHANGELOG_PENDING.md]: Fixed in the force-pushed branch.

**Pull request merged**